cmdstanpy/lib.py

```diff
@@ -527,27 +552,59 @@ def sample(self) -> np.ndarray:
+            self.parse_sample()
         return self._sample
 
-    def get_sample(self) -> np.ndarray:
+    def parse_sample(self) -> None:
         """
-        Returns posterior sample.
-        The first time this function is called it assembles the sample
-        from the stan_csv files; subsequent calls to this function
-        return the assembled sample.
+        Allocates and populates the stepsize, metric, and drawset arrays
+        by parsing the validated stan_csv files if they don't already exist.
         """
-        sample = np.empty(
+        if not (self._stepsize is None and self._metric is None and
+                    self._sample is None):
+            return
+        self._stepsize = np.empty(self._chains, dtype=float)
+        if self._metric_type == 'diag_e':
+            self._metric = np.empty(
+                (self._chains, self._num_params), dtype=float)
+        else:
+            self._metric = np.empty(
+                (self._chains, self._num_params, self._num_params),
+                dtype=float)
+        self._sample = np.empty(
             (self._draws, self._chains, len(self._column_names)), dtype=float,
             order='F'
         )
         for chain in range(self._chains):
-            draw = 0
-            with open(self._csv_files[chain], 'r') as fd:
-                for line in fd:
-                    if line.startswith('#') or line.startswith('lp__,'):
-                        continue
-                    vs = [float(x) for x in line.split(',')]
-                    sample[draw, chain, :] = vs
-                    draw += 1
-        return sample
+            try:
+                fp = open(self._csv_files[chain], 'r')
+                try:
+                    # read past initial comments, column header
+                    line = fp.readline().strip()
+                    while len(line) > 0 and line.startswith('#'):
+                        line = fp.readline().strip()
+                    line = fp.readline().strip()  # adaptation header
+                    # stepsize
+                    line = fp.readline().strip()
+                    label, stepsize = line.split('=')
+                    self._stepsize[chain] = float(stepsize.strip())
+                    line = fp.readline().strip()  # metric header
+                    # metric
+                    if self._metric_type == 'diag_e':
+                        line = fp.readline().lstrip(' #\t')
+                        xs = line.split(',')
+                        self._metric[chain, :] = [float(x) for x in xs]
+                    else:
+                        for i in range(self._num_params):
+                            line = fp.readline().lstrip(' #\t')
+                            xs = line.split(',')
+                            self._metric[chain, i, :] = [float(x) for x in xs]
+                    # draws
+                    for i in range(self._draws):
+                        line = fp.readline().lstrip(' #\t')
```

Comment:
What comments can be found in sample? If they are not needed could this loop be replaced with `np.loadtxt` or `np.genfromtxt`?

Reply:
the timing information is written out as a comment following the draws.
the `stan_csv` format is an abomination and we must replace it.  until then, brute-force parsing.